### ISSUE TITLE
Best practice - Azure Function AAD with sample

### ISSUE THREAD

**Issue by diberry** [AzureAD/microsoft-authentication-library-for-js#4168]
### Core Library

Passport Azure AD (passport-azure-ad)

### Core Library Version

^4.3.0

### Wrapper Library

Not Applicable

### Wrapper Library Version

None

### Description

I'm not sure which library so please ignore the selections. If I have a React app  passing a bearer token (after they login with MSAL-React) for the user to an Azure Function, what libraries and functionality can I use to rehydrate that user's auth from an AAD token to an Azure credential I can use to connect to the Azure SDKs for JS? 

I found this [sample](https://github.com/Azure-Samples/ms-identity-nodejs-webapi-azurefunctions/blob/master/Function/MyHttpTrigger/index.js) but it is a year old and doesn't seem very function-ish. 

### MSAL Configuration

```javascript
{
    identityMetadata: `https://${auth.authority}/${auth.tenantID}/${auth.version}/${auth.discovery}`,
    issuer: `https://${auth.authority}/${auth.tenantID}/${auth.version}`,
    clientID: auth.clientID,
    audience: auth.audience,
    validateIssuer: auth.validateIssuer,
    passReqToCallback: auth.passReqToCallback,
    loggingLevel: auth.loggingLevel,
    scope: auth.scope
}
```

### Relevant Code Snippets

```javascript
const express = require('express');
const passport = require('passport');
const auth = require('../auth.json');

const createHandler = require('azure-function-express').createHandler;

const BearerStrategy = require("passport-azure-ad").BearerStrategy;

const options = {
    identityMetadata: `https://${auth.authority}/${auth.tenantID}/${auth.version}/${auth.discovery}`,
    issuer: `https://${auth.authority}/${auth.tenantID}/${auth.version}`,
    clientID: auth.clientID,
    audience: auth.audience,
    validateIssuer: auth.validateIssuer,
    passReqToCallback: auth.passReqToCallback,
    loggingLevel: auth.loggingLevel,
    scope: auth.scope
};

const bearerStrategy = new BearerStrategy(options, (token, done) => {
    // Send user info using the second argument
    done(null, {}, token);
});

const app = express();

app.use(require('morgan')('combined'));

app.use(require('body-parser').urlencoded({ 'extended': true }));

app.use(passport.initialize());

passport.use(bearerStrategy);

// Enable CORS (for local testing only -remove in production/deployment)
app.use((req, res, next) => {
    res.header('Access-Control-Allow-Origin', '*');
    res.header('Access-Control-Allow-Headers', 'Authorization, Origin, X-Requested-With, Content-Type, Accept');
    next();
});

// Expose and protect API endpoint
app.get('/api', passport.authenticate('oauth-bearer', { session: false }),
    (req, res) => {
        console.log('Validated claims: ', req.authInfo);

        // Service relies on the name claim.  
        res.status(200).json({
            'name': req.authInfo['name'],
            'issued-by': req.authInfo['iss'],
            'issued-for': req.authInfo['aud'],
            'using-scope': req.authInfo['scp']
        });
    }
);

module.exports = createHandler(app);
```

### Identity Provider

Azure AD / MSA

### Source

Internal (Microsoft)

**Comment by jasonnutter** [comment]
@diberry Typically for a scenario like this I would recommend using the On Behalf Of flow in MSAL Node, as shown here: https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/samples/msal-node-samples/on-behalf-of

TLDR:
1. SPA requests an access token for the custom scope setup for the Azure Function, and gets consent for any downstream resources needed by the Azure Function.
2. SPA sends access token for custom scope to Azure Function, which uses it in an OBO flow to get an access token for the downstream resource.
3. Azure Function uses second access token against downstream resource to get needed data.

There is no sample currently for Azure Functions, unfortunately, but is something we can add to our backlog.

Unfortunately, it does not appear that the Azure SDK has a OBO credential provider, @sadasant can you confirm?

**Comment by sadasant** [comment]
Last week we released Identity v2, which includes an OnBehalfOfCredential. Here is the simplest sample I could think of regarding how this credential can be used: https://github.com/Azure/azure-sdk-for-js/blob/main/sdk/identity/identity/samples/AzureIdentityExamples.md#authenticate-on-behalf-of

It does use msal-node internally 🙂

**Comment by jasonnutter** [comment]
> Last week we released Identity v2, which includes an OnBehalfOfCredential. Here is the simplest sample I could think of regarding how this credential can be used: https://github.com/Azure/azure-sdk-for-js/blob/main/sdk/identity/identity/samples/AzureIdentityExamples.md#authenticate-on-behalf-of
> 
> It does use msal-node internally 🙂

Great, thanks! @diberry I would recommend looking at the OBO flow, and let us know if you have questions about that.

**Comment by diberry** [comment]
@sadasant @jasonnutter Thank you for your help with this. I think the OBO sample that Daniel points out is for a client (interactive) user and doesn't fit my need for an Azure Function.

My Azure Function gets the user's accesstoken and needs to create the Azure Credential. Looking for the descriptions, I think that would be [AuthorizationCodeCredential](https://www.npmjs.com/package/@azure/identity#credential-classes) but that uses an authorization code - passed from the client - or it that the token?

I'm looking for the flow and code that demonstrates OBO on the server (serverless Azure Function) so that the resulting credential can be used to access Azure from the Azure SDKs (not REST). 

Are Authorization code flow and OBO separate in that sense that they don't combine and one is better than the other in my scenario? 

How is the resulting credential of auth code different from OBO? They seem kinda the same from a certain perspective.

**Comment by sadasant** [comment]
@diberry let me ask you something, so I can understand this.

Would this satisfy the scenario you’re presenting to us?

1. Authenticate within the browser (with InteractiveBrowserCredential or msal-browser).
2. Grab the resulting token into a variable.
3. Send a POST request from single-page application to a web server with the resulting token in the body.
4. In the backend, to instantiate a credential using the token pre-fetched from the web, so that this credential won’t need to get new tokens.
5. To use that instance of a credential with the existing Azure SDKs.

Is that the idea?

We have a sample on how to create a new credential with a pre-fetched token: https://github.com/Azure/azure-sdk-for-js/blob/main/sdk/identity/identity/samples/AzureIdentityExamples.md#authenticating-with-a-pre-fetched-access-token

In summary, a TokenCredential can just be an object with a getToken function that returns the pre-fetched token, so a new one can be something like:

```ts
// This would be in the Node.js server, after the web app has sent a request with the access token in the body.

// Just the access token received from the web call.
// Keep in mind that an “AccessToken” for the Azure SDKs is an object with { token, expiresOnTimestamp }
// You can see this type here: https://github.com/Azure/azure-sdk-for-js/blob/main/sdk/core/core-auth/src/tokenCredential.ts#L60-L70
const webAccessToken = getWebAccessToken();

const credential = {
  async getToken() {
    return webAccessToken
  }
};

// Now we would use the Azure SDKs…
const client = new KeyClient(vaultUri, credential);
```

Would that help? 🤔 feedback appreciated! 🙏 🙇‍♂️

**Comment by diberry** [comment]
@sadasant This is great - thank you. I believe the access token should be passed to the function in the auth header, instead of a Post body. Then in the Azure Function API, the auth header holding that bearer token needs to be parsed to get back to just the access token. Then you mention that the rehydration process also needs the time stamp. I think that is the "exp" property from the access token object returned from AAD. Ideally, is the bearer token's expiration passed in via header? Is there anything on the Azure AAD side that validates that token with the expiration value used in hydration? Meaning could I send in a value that indicates the token never expires in the Azure Function API, that is incorrect/diff from the exp received originally with the token on the React Client?

**Comment by sadasant** [comment]
> I believe the access token should be passed to the function in the auth header, instead of a Post body. 

That sounds good! Sorry, I was mainly trying to think on how it would work.

>  Ideally, is the bearer token's expiration passed in via header? Is there anything on the Azure AAD side that validates that token with the expiration value used in hydration? […]

Oh, I have good news for you. The expiration date is only used on the SDK client side, not by Azure AD. The expiration date tells our HTTP pipelines when to consider the token expired. The service does this by themselves, of course. They don’t rely on us sending this expiration date.

On our HTTP pipeline, what gets used to send authenticated requests is only the token, not the expiration date.

Though if you create your own credential, then the HTTP pipeline would need the expiration date in order to attempt to renew the token before it expires. Token renewal happens internally by calling to `getToken` again, so if `getToken` always returns the same access token, it won’t be possible for the HTTP pipeline to get a new valid token by itself.

In your scenario, since the token is retrieved from the browser’s single-page application, the HTTP pipeline won’t be able to auto-renew the token. This means that, for your use case, the expiration date could be any date in the future. You could choose to avoid further network requests by comparing the expiration date you might have in the backend before making SDK calls. More complex ideas are available, of course.

In summary: Expiration date only matters to know when to send network requests. AAD does not receive expiration dates through headers, because they keep track of the real expiration internally, of course.

I hope that helps. Please let me know if there’s anything else I could do 🙇‍♂️

**Comment by diberry** [comment]
Also, would this still be OBO? Or can a rehydrate in any flow?

**Comment by sadasant** [comment]
@diberry this is not OBO, since the user token doesn’t get used to authenticate on behalf of another identity.

(OBO also has a bunch of weird special extra steps to configure)

**Comment by jasonnutter** [comment]
@diberry Are your questions answered here?

**Comment by diberry** [comment]
@jasonnutter @sadasant Handling this offline.
